queries/haskell/highlights.scm

```diff
@@ -190,26 +176,44 @@
          ]))
      (quasiquote)
      ((exp_name) . (operator))
-    ]))
+    ])+
+ (#eq? @variable @_name))
+
+((signature name: (variable) @variable)
+ . (comment)*
+ . (function 
+    name: (variable) @_name
+    rhs: (exp_infix 
+      [
+       (exp_literal)
+       (exp_apply 
+         (exp_name 
+           [(constructor)
+            (variable)
+           ]))
+       (quasiquote)
+       ((exp_name) . (operator))
+      ]))+
  (#eq? @variable @_name))
 
 ((signature name: (variable) @function)
- (function 
-  name: (variable) @_name
-  patterns: (patterns))
+ . (comment)*
+ . (function 
+    name: (variable) @_name
```

Comment:
This query wont be working very well right now for neovim as it's captures inside quantifiers. What's the result looking like while using helix (preferably helix nightly) with this query?

Reply:
This query isn't in Helix. I don't think I've actually encountered a case where this is necessary (yet). I just added it for better reliability.
The reason is as follows:

- In Haskell, everything is a function. So if you have `x = 5`, `x` is a function that evaluates to 5.
- The treesitter parser will parse `x` as a `variable` in  `foo = x`, but `foo` a function, and there are highlights that try to treat `foo` as a variable.
- There are some ways to detect this (e.g. if the lhs has parameters), and the tree-sitter queries are quite good at this.

A sure way to detect this is if there's an accompanying signature of `type: fun`.
For example:

```haskell
foo :: Int -> Int
foo = x
```

The problem is, signatures and functions are at the same level in the tree.
This makes sense, because in Haskell, signatures can be defined anywhere in the same module.
So something like

```haskell
foo :: Int -> Int

bar = 5

foo = x
```

is perfectly valid (although an unlikely use case).
That's why I had all those unanchored queries.

The query you commented on tries to match things like

```haskell
foo :: Int -> Int
-- some comment
foo 5 = 0
-- another comment
foo x = x + 1
```

But now that I'm writing this, I think there's no need for this.
The only case we need this for is 

```haskell
foo :: Int -> Int
-- some comment
-- maybe another comment
foo = x
```

because 

```haskell
foo :: Int -> Int
foo = 5
foo = x
```
is invalid haskell

and

```haskell
foo x = --<...>
```

is easy to capture without having to know about signatures.

Reply:
I just realised there were a bunch of queries that tested for this. They were just missing `. (comment)* .` in between the `(signature) . (function)` anchors (see my second commit).

Reply:
I don't think having `(comment)*` here will be nice. Treesitter should not be responsible for this level of granular control. It should be left to the LSP.

Just `(signature)` and `(function)` that sits alongside one another will be better imo

Reply:
It's quite common to have comments between the signature and the function (this is the case in the test file, too).
The language server doesn't support highlighting hints yet.

Reply:
I'm somewhat afraid for matching this kind of pattern, mostly because of the double quantifiers

something like 
`(signature) . (some 10 or more comment lines) . (10 consecutive functions)` can be quite costly

Reply:
Also, I'm pretty sure the original intention is not to match that far away. I don't think you need to capture all functions going right after a signature. I believe matching a function right after the signature is fine. Trying too much to make the highlights work well won't be worth it